File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        Arrays.sort(potions);
        int[] searchValues = new int[spells.length];
        for (int i = 0; i < spells.length; i++) {
            searchValues[i] = (int) Math.ceil((double) success / spells[i]);
        }
        int[] pairs = new int[spells.length];
        for (int i = 0; i < spells.length; i++) {
            int count = potions.length - binarySearch(potions, searchValues[i]);
            pairs[i] = count;
        }
        return pairs;
    }

    private int binarySearch(int[] array, int search) {
        int low = 0;
        int high = array.length - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (array[mid] < search) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return low;
    }
}

File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.kt
class Solution {
    fun successfulPairs(spells: IntArray, potions: IntArray, success: Long): IntArray {
        val sortedPotions = potions.sortedArray()
        val searchValues = spells.map { ceil(success.toDouble() / it).toInt() }
        val pairs = IntArray(spells.size)

        for ((index, searchVal) in searchValues.withIndex()) {
            val count = sortedPotions.size - binarySearch(sortedPotions, searchVal)
            pairs[index] = count
        }
        return pairs
    }

    private fun binarySearch(array: IntArray, search: Int): Int {
        var low = 0
        var high = array.size - 1
        while (low <= high) {
            val mid = low + (high - low) / 2
            if (array[mid] < search) {
                low = mid + 1
            } else {
                high = mid - 1
            }
        }
        return low
    }
}
